frontend/src/components/resourceMap/graph/graphGrouping.tsx

```diff
@@ -126,18 +126,49 @@ const getConnectedComponents = (nodes: GraphNode[], edges: GraphEdge[]): GraphNo
 };
 
 /**
- * Try to find a "main" node in the workload group
- * If can't find anything return the first node
+ * Try to find a "main" node in the group based on weight.
+ * Higher weight nodes are considered more important.
+ * If weights are equal, falls back to the hardcoded hierarchy.
+ *
+ * @param nodes - Array of nodes to find the main node from
+ * @returns The node with the highest weight (most important), or undefined if array is empty
  */
-export const getMainNode = (nodes: GraphNode[]) => {
-  const deployment = nodes.find(it => it.kubeObject?.kind === 'Deployment');
-  const replicaSet = nodes.find(it => it.kubeObject?.kind === 'ReplicaSet');
-  const daemonSet = nodes.find(it => it.kubeObject?.kind === 'DaemonSet');
-  const statefulSet = nodes.find(it => it.kubeObject?.kind === 'StatefulSet');
-  const job = nodes.find(it => it.kubeObject?.kind === 'Job');
-  const cronJob = nodes.find(it => it.kubeObject?.kind === 'CronJob');
-
-  return deployment ?? replicaSet ?? daemonSet ?? statefulSet ?? cronJob ?? job ?? nodes[0];
+export const getMainNode = (nodes: GraphNode[]): GraphNode | undefined => {
+  if (nodes.length === 0) {
+    return undefined;
+  }
+
+  if (nodes.length === 1) {
+    return nodes[0];
+  }
+
+  // sort nodes by weight (descending) and return the highest weighted node
+  const sortedByWeight = [...nodes].sort((a, b) => {
```

Comment:
Could you rewrite this function so that it doesn't create a copy and doesn't sort them all? It's possible to find a node with the biggest weight in linear time without allocating a new array